mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -34,33 +35,142 @@
-
+    auto calleeOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+
+    func::FuncOp fnFoldedOp;
+
+    if (!calleeOp->hasAttr("qnode")) {
+        // Traverse the callgraph, copy all the function to a `.zne` version and fold qnodes
+        traverseCallGraph(calleeOp, /*symbolTable=*/nullptr, [&](func::FuncOp funcOp) {
+            if (!funcOp->hasAttr("qnode")) {
+                PatternRewriter::InsertionGuard insertGuard(rewriter);
+                rewriter.setInsertionPointToStart(moduleOp.getBody());
+
+                // Copy the function and create a .zne counter part and add the scale factor as last
+                // argument
+                auto currentFnFoldedOp = createZneFunc(funcOp, rewriter);
+                // Folding of the qnodes and replace their calls with the folded version
+                if (containsQnodes(currentFnFoldedOp)) {
+                    currentFnFoldedOp.walk([&](func::CallOp callOp) {
+                        PatternRewriter::InsertionGuard insertGuard(rewriter);
+                        func::FuncOp funcOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            op, callOp.getCalleeAttr());
+                        std::string foldedName = callOp.getCalleeAttrName().str() + ".folded";
+                        func::FuncOp foldedOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            moduleOp, rewriter.getStringAttr(foldedName));
+                        if (funcOp->hasAttr("qnode") and !foldedOp) {
+                            // Create the folded circuit function
+                            auto foldedCircuitAttr =
+                                getOrInsertFoldedCircuit(loc, rewriter, funcOp, foldingAlgorithm);
+                            func::FuncOp foldedOp =
+                                SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                                    moduleOp, foldedCircuitAttr);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.setInsertionPoint(callOp);
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
+                        }
+                        else if (funcOp->hasAttr("qnode") and foldedOp) {
+                            rewriter.setInsertionPoint(callOp);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
```

Comment:
These two blocks look identical, maybe you can simplify the if statement by factoring them out?

Reply:
Simplified :+1: